File: src/utils/stats.ts
// Reading statistics derived from history.json (all local, no network).
// History records the latest position per title, so figures reflect progress
// reached rather than cumulative re-reads.

import type { HistoryEntry } from "./history.ts";

export interface Stats {
  titles: number;
  chaptersProgressed: number;
  finished: number;
  inProgress: number;
  topTitles: Array<{ title: string; chapters: number; total: number; pct: number }>;
  weekday: number[]; // length 7, Sun..Sat — count of titles last read that day
  activeDays: number;
  longestStreak: number;
  currentStreak: number;
  since?: string;
  lastReadTitle?: string;
}

function dayKey(iso: string): string {
  const d = new Date(iso);
  return `${d.getFullYear()}-${d.getMonth() + 1}-${d.getDate()}`;
}

function streaks(days: Set<string>): { longest: number; current: number } {
  if (days.size === 0) return { longest: 0, current: 0 };
  const times = [...days]
    .map((k) => {
      const [y, m, d] = k.split("-").map(Number);
      return new Date(y, m - 1, d).getTime();
    })
    .sort((a, b) => a - b);
  const DAY = 86_400_000;
  let longest = 1;
  let run = 1;
  for (let i = 1; i < times.length; i++) {
    const gap = Math.round((times[i] - times[i - 1]) / DAY);
    run = gap === 1 ? run + 1 : 1;
    longest = Math.max(longest, run);
  }
  // current streak: consecutive days ending today or yesterday
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  let current = 0;
  let cursor = today.getTime();
  const has = (t: number): boolean => days.has(dayKey(new Date(t).toISOString()));
  if (!has(cursor)) cursor -= DAY; // allow "yesterday" to count
  while (has(cursor)) {
    current++;
    cursor -= DAY;
  }
  return { longest, current };
}

export function computeStats(history: HistoryEntry[]): Stats {
  const titles = history.length;
  let chaptersProgressed = 0;
  let finished = 0;
  let inProgress = 0;
  const weekday = [0, 0, 0, 0, 0, 0, 0];
  const days = new Set<string>();
  let since: string | undefined;

  for (const h of history) {
    const reached = Math.max(h.lastChapterIndex + 1, h.lastChapterNumber || 0);
    chaptersProgressed += Math.max(0, h.lastChapterIndex + 1);
    if (h.totalChapters > 0 && reached >= h.totalChapters) finished++;
    else inProgress++;
    const d = new Date(h.lastReadAt);
    if (!Number.isNaN(d.getTime())) {
      weekday[d.getDay()]++;
      days.add(dayKey(h.lastReadAt));
      if (!since || h.lastReadAt < since) since = h.lastReadAt;
    }
  }

  const topTitles = [...history]
    .map((h) => {
      const chapters = Math.max(h.lastChapterIndex + 1, 0);
      const total = h.totalChapters || chapters;
      return { title: h.title, chapters, total, pct: total ? Math.round((chapters / total) * 100) : 0 };
    })
    .sort((a, b) => b.chapters - a.chapters)
    .slice(0, 8);

  const { longest, current } = streaks(days);

  return {
    titles,
    chaptersProgressed,
    finished,
    inProgress,
    topTitles,
    weekday,
    activeDays: days.size,
    longestStreak: longest,
    currentStreak: current,
    since,
    lastReadTitle: history[0]?.title,
  };
}

File: go-rewrite/internal/util/stats.go
// Reading statistics derived from history.json (all local, no network).

package util

import (
	"sort"
	"time"
)

type TopTitle struct {
	Title    string
	Chapters int
	Total    int
	Pct      int
}

type Stats struct {
	Titles             int
	ChaptersProgressed int
	Finished           int
	InProgress         int
	TopTitles          []TopTitle
	Weekday            [7]int // Sun..Sat
	ActiveDays         int
	LongestStreak      int
	CurrentStreak      int
	Since              string
	LastReadTitle      string
}

func dayKey(t time.Time) string { return t.Local().Format("2006-01-02") }

func ComputeStats(history []HistoryEntry) Stats {
	var s Stats
	s.Titles = len(history)
	days := map[string]bool{}
	for _, h := range history {
		reached := max(h.LastChapterIndex+1, int(h.LastChapterNumber))
		s.ChaptersProgressed += max(0, h.LastChapterIndex+1)
		if h.TotalChapters > 0 && reached >= h.TotalChapters {
			s.Finished++
		} else {
			s.InProgress++
		}
		if t, err := time.Parse(time.RFC3339, h.LastReadAt); err == nil {
			s.Weekday[int(t.Local().Weekday())]++
			days[dayKey(t)] = true
			if s.Since == "" || h.LastReadAt < s.Since {
				s.Since = h.LastReadAt
			}
		}
	}
	for _, h := range history {
		ch := max(h.LastChapterIndex+1, 0)
		total := h.TotalChapters
		if total == 0 {
			total = ch
		}
		pct := 0
		if total > 0 {
			pct = ch * 100 / total
		}
		s.TopTitles = append(s.TopTitles, TopTitle{h.Title, ch, total, pct})
	}
	sort.SliceStable(s.TopTitles, func(a, b int) bool {
		return s.TopTitles[a].Chapters > s.TopTitles[b].Chapters
	})
	if len(s.TopTitles) > 8 {
		s.TopTitles = s.TopTitles[:8]
	}
	s.ActiveDays = len(days)
	s.LongestStreak, s.CurrentStreak = streaks(days)
	if len(history) > 0 {
		s.LastReadTitle = history[0].Title
	}
	return s
}

func streaks(days map[string]bool) (longest, current int) {
	if len(days) == 0 {
		return 0, 0
	}
	var times []time.Time
	for k := range days {
		if t, err := time.ParseInLocation("2006-01-02", k, time.Local); err == nil {
			times = append(times, t)
		}
	}
	sort.Slice(times, func(a, b int) bool { return times[a].Before(times[b]) })
	longest, run := 1, 1
	for i := 1; i < len(times); i++ {
		if int(times[i].Sub(times[i-1]).Hours()/24+0.5) == 1 {
			run++
		} else {
			run = 1
		}
		longest = max(longest, run)
	}
	// current streak: consecutive days ending today or yesterday
	cursor := time.Now().Local().Truncate(24 * time.Hour)
	if !days[dayKey(cursor)] {
		cursor = cursor.AddDate(0, 0, -1)
	}
	for days[dayKey(cursor)] {
		current++
		cursor = cursor.AddDate(0, 0, -1)
	}
	return longest, current
}
